### PR TITLE
[http] add pre-emptive basic authentication and fix header handling

### LINKED ISSUE #9567: [HTTP] Basic auth not working when configured in thing
Problem: Basic Auth seems to be not working when filling username/password in the UI thing configuration

Version: OpenHAB 3.0 Release

Created HTTP thing with one channel to query a REST API with a GET request:
(please ignore the additonal headers, I added them later on to get it going as a workaround)
```
UID: http:url:openems-rest
label: OpenEMS
thingTypeUID: http:url
configuration:
  authMode: BASIC
  ignoreSSLErrors: false
  baseURL: http://examplehost:8084/rest/channel/
  password: admin
  refresh: 30
  commandMethod: PUT
  contentType: application/json
  timeout: 3000
  username: admin
  bufferSize: 2048
  headers:
    - Authorization=Basic YWRtaW46YWRtaW4=
channels:
  - id: meter0ActivePower
    channelTypeUID: http:string
    label: meter0ActivePower
    description: ""
    configuration:
      mode: READONLY
      stateExtension: meter0/ActivePower
```
This happens:
```
20:28:14.396 [TRACE] [http.internal.http.RefreshingUrlCache] - Sending to 'http://kube.gruning.eu:8084/rest/channel/meter0/ActivePower': Method = {GET}, Headers = {Accept-Encoding: gzip, User-Agent: Jetty/9.4.20.v20190813}, Content = {null}
20:28:14.397 [TRACE] [tp.internal.http.HttpResponseListener] - Received from 'http://kube.gruning.eu:8084/rest/channel/meter0/ActivePower': Code = {400}, Headers = {Date: Mon, 28 Dec 2020 19:28:14 GMT, Content-Type: application/json, Content-Length: 127, Server: Jetty(9.4.32.v20200930)}, Content = {{"jsonrpc":"2.0","id":"00000000-0000-0000-0000-000000000000","error":{"code":1003,"message":"Authentication failed","data":[]}}}
20:28:14.398 [WARN ] [tp.internal.http.HttpResponseListener] - Requesting 'http://kube.gruning.eu:8084/rest/channel/meter0/ActivePower' (method='GET', content='null') failed: 400 Bad Request
```

Then I manually added the authorization header which then leads to the expected result:
```
20:29:31.162 [TRACE] [http.internal.http.RefreshingUrlCache] - Sending to 'http://kube.gruning.eu:8084/rest/channel/meter0/ActivePower': Method = {GET}, Headers = {Accept-Encoding: gzip, User-Agent: Jetty/9.4.20.v20190813, Authorization: Basic YWRtaW46YWRtaW4=}, Content = {null}
20:29:31.166 [TRACE] [tp.internal.http.HttpResponseListener] - Received from 'http://kube.gruning.eu:8084/rest/channel/meter0/ActivePower': Code = {200}, Headers = {Date: Mon, 28 Dec 2020 19:29:31 GMT, Content-Type: application/json, Content-Length: 48, Server: Jetty(9.4.32.v20200930)}, Content = {{"value":1427,"type":"integer","writable":false}}
20:29:31.167 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'OpenEMS_meter0ActivePower' changed from NULL to {"value":1427,"type":"integer","writable":false}
```

To me it looks like there is no authorization header in the request at all (despite config in the UI). Looks like a bug to me?

### PR DESCRIPTION
**Pull request by J-N-K**
Fixes #9567 

In some cases HTTP servers  respond with 400/Bad Request (and a message saying "unauthorized" in the content) instead of 401/Unauthorized if they require authentication. Since Jetty sends auth-headers only if the are requested by the server via 401, connections to these servers fail. This allows to force sending the Basic-Auth header in the initial request.

Also made sure that empty additional header entries are removed before headers are processed.

### CODE REVIEW

File: bundles/org.openhab.binding.http/README.md [hunk 1]
```diff
@@ -18,14 +18,19 @@ It can be extended with different channels.
...
 | `commandMethod`   | no       |   GET   | Method used for sending commands `GET`, `PUT`, `POST`. |
 | `contentType`     | yes      |    -    | MIME content-type of the command requests. Only used for  `PUT` and `POST`. |
 | `encoding`        | yes      |    -    | Encoding to be used if no encoding is found in responses (advanced parameter). |  
 | `headers`         | yes      |    -    | Additional headers that are sent along with the request. Format is "header=value".| 
 | `ignoreSSLErrors` | no       |  false  | If set to true ignores invalid SSL certificate errors. This is potentially dangerous.|
 
-*Note:* optional "no" means that you have to configure a value unless a default is provided and you are ok with that setting.
+*Note:* Optional "no" means that you have to configure a value unless a default is provided and you are ok with that setting.
+
+*Note:* The `BASIC_PREEMPTIVE` modes adds basic authentication headers even if the server did not request authentication.
```
Comment (t2000) [inline bundles/org.openhab.binding.http/README.md:30]: Typo fix:

```suggestion
*Note:* The `BASIC_PREEMPTIVE` mode adds basic authentication headers even if the server did not request authentication.
```

**Comment by fwolter** [comment]
Can you fix thr DCO?

**Review by fwolter — APPROVED** [review]
LGTM

**Comment by J-N-K** [comment]
DCO was broken because I used "commit suggestion". Overrided.

**Pull request merged**